torchrl/envs/libs/meltingpot.py

```diff
@@ -0,0 +1,594 @@
+            group_action_np = self.full_action_spec[group, "action"].to_numpy(
+                group_action
+            )
+            for index, agent in enumerate(agents):
+                action_dict[agent] = group_action_np[index]
+
+        actions = [action_dict[agent] for agent in self.agent_names]
+        timestep = self._env.step(actions)
+        self.num_cycles += 1
+
+        rewards = timestep.reward
+        done = timestep.last() or (
+            (self.num_cycles >= self.max_steps) if self.max_steps is not None else False
+        )
+        obs = timestep.observation
+
+        td = TensorDict(
+            {
+                "done": self.full_done_spec["done"].encode(done),
+                "terminated": self.full_done_spec["terminated"].encode(done),
+            },
+            batch_size=self.batch_size,
+        )
+        # Global state
+        td.update(_filter_global_state_from_dict(obs[0], world=True))
+
+        for group, agent_names in self.group_map.items():
+            agent_tds = []
+            for index_in_group, agent_name in enumerate(agent_names):
+                global_index = self.agent_names_to_indices_map[agent_name]
+                agent_obs = self.observation_spec[group, "observation"][
+                    index_in_group
+                ].encode(_filter_global_state_from_dict(obs[global_index], world=False))
+                agent_reward = self.full_reward_spec[group, "reward"][
+                    index_in_group
+                ].encode(rewards[global_index])
+                agent_td = TensorDict(
+                    source={
+                        "observation": agent_obs,
+                        "reward": agent_reward,
+                    },
+                    batch_size=self.batch_size,
+                    device=self.device,
+                )
+
+                agent_tds.append(agent_td)
+            agent_tds = torch.stack(agent_tds, dim=0)
+            td.update({group: agent_tds})
+
+        return td
+
+    def render(self, mode="human", filename=None) -> np.ndarray:
+        """Renders the environment using matplotlib.
+
+        Args:
+            mode (str, optional): One of ``"human"``, ``"rgb_array"``. If ``"human"`` it renders the
+                environment in the GUI. In any case the function returns a RGB array. Defaults to ``"human"``
+            filename (str, optional): Filename to save the render to. Defaults to ``None``,
+                in which case no file is saved.
+
+        Returns:
+            np.ndarray: The rendered image
+
+        """
+        from matplotlib import pyplot as plt
+
+        rgb_arr = self._env.observation()[0]["WORLD.RGB"]
+        if mode == "human":
+            plt.cla()
+            plt.imshow(rgb_arr, interpolation="nearest")
+            if filename is None:
+                plt.show(block=False)
+            else:
+                plt.savefig(filename)
+        return rgb_arr
```

Comment:
the torchrl way of doing this is to put the rendered images in a "pixels" entry in the output tensordict and a `from_pixels=True` arg in constructor. So far, no env supports render by design.

Reply:
Meltingpot has rgb images by default part of the observation dictionary under the keys : "WORLD.RGB" and "RGB".

This is just a util to show those to the GUI or save them to file.

I thought it would be helpful since MeltingPot itself does not have it and i need to call `env.render()` in BenchMARL, but if you think it is not helpful i'll move it to BenchMARL.

The torchrl way seems to be related to input output. This function is not and can be called at any time

Reply:
Oh I get it you're not rendering anything but just using matplotlib to display it.

I don't think that should be within this env, because that makes the whole class dependent on matplotlib and that doesn't seem super neat (one more extra dep to handle etc).

Reply:
matplotlib is imported only if you call the function.

Which seems fair to me no?

Reply:
actually i can make it so that matplotlib is imported only is the users calls the function and asks for GUI

Reply:
Sure, but we need to test it if we integrate it no?
Plus that seems fairly simple and orthogonal to what the env is doing (not required to run it).
To me that shouldn't be part of the env wrapper

Reply:
I think that prividing the method done right is a plus for users so I do not really agree.

Anyway I'll just put it in BenchMARL

Reply:
what do you think about a method called `get_rgb_image`, which returns an image of the env at any time?

Reply:
yeah sure that sounds reasonable to me!
the thing I was skeptical about was the matplotlib dependency.